### PR TITLE
authentication: Update passkey auth to run the command after auth of the user

### PR DESCRIPTION
**Pull request by madhuriupadhye**
Update the passkey authentication method to run command after authentication of user
and get the expected output of that command.

### CODE REVIEW

File: sssd_test_framework/utils/authentication.py [hunk 1]
```diff
@@ -397,44 +407,82 @@ def passkey(self, username: str, *, pin: str | int, device: str, ioctl: str, scr
...
+                exit $code
+            }}
+
             # It takes some time to get authentication failure
             set timeout {DEFAULT_AUTHENTICATION_TIMEOUT}
             set prompt "\n.*\[#\$>\] $"
+            set command "{command}"
 
-            spawn "{playback_umockdev}"
+            spawn {playback_umockdev}
```
Comment (spoore1) [inline sssd_test_framework/utils/authentication.py:435]: You can probably add the quotes back here.   I removed them when troubleshooting an issue but, forgot to put them back.

**Review by spoore1 — CHANGES_REQUESTED** [review]
Just a couple of small things to update.

File: sssd_test_framework/utils/authentication.py [hunk 2]
```diff
@@ -397,44 +407,82 @@ def passkey(self, username: str, *, pin: str | int, device: str, ioctl: str, scr
...
-        return result.rc == 0
+        expect_data = result.stdout_lines[-3:]
+
+        # Get command exit code.
+        cmdrc = int(expect_data[2].split(":")[1].strip())
+
+        # Alter stdout, first line is spawned command, the last three are our expect output.
+        stdout = "\n".join(result.stdout_lines[1:-3])
+
+        return result.rc, cmdrc, stdout, result.stderr
```
Comment (spoore1) [inline sssd_test_framework/utils/authentication.py:485]: by changing the return,  you'll need to make sure you change the code that uses this method.

File: sssd_test_framework/utils/authentication.py [hunk 3]
```diff
@@ -340,7 +340,17 @@ def password_expired(self, username: str, password: str, new_password: str) -> b
...
+    def passkey(
+        self,
+        username: str,
+        *,
+        pin: str | int,
+        device: str,
+        ioctl: str,
+        script: str,
+        command: str | '',
+        expected_output: str | ''
```
[the remark this answers is not in the document]
Comment (madhuriupadhye) [reply sssd_test_framework/utils/authentication.py]: updated.

File: sssd_test_framework/utils/authentication.py (hunk 2 above)
Comment (madhuriupadhye) [reply sssd_test_framework/utils/authentication.py:486]: We need this to get console messages.

File: sssd_test_framework/utils/authentication.py (hunk 1 above)
Comment (madhuriupadhye) [reply sssd_test_framework/utils/authentication.py]: updated

File: requirements.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
 jc
 pytest
 python-ldap
-pytest-mh >= 1.0.11
+#pytest-mh >= 1.0.11
```
Comment (spoore1) [inline requirements.txt:4]: Was this commented out for debugging?

File: sssd_test_framework/utils/authentication.py [hunk 5]
```diff
@@ -340,7 +340,17 @@ def password_expired(self, username: str, password: str, new_password: str) -> b
...
+    def passkey(
+        self,
+        username: str,
+        *,
+        pin: str | int,
+        device: str,
+        ioctl: str,
+        script: str,
+        command: str | Any,
+        expected_output_from_cmd: str | Any
```
Comment (spoore1) [inline sssd_test_framework/utils/authentication.py:352]: Is "Any" better here than "None = ''"?

Like we discussed earlier, the issue you found when `command` is defined and `expected_output_from_cmd` is set to an empty string, may require special handling.   Just something to consider.

**Review by spoore1 — COMMENTED** [review]
Just a few small questions.

File: requirements.txt (hunk 4 above)
Comment (madhuriupadhye) [reply requirements.txt:4]: Its just for testing.

File: sssd_test_framework/utils/authentication.py (hunk 5 above)
Comment (madhuriupadhye) [reply sssd_test_framework/utils/authentication.py]: updated the code.

File: sssd_test_framework/utils/authentication.py [hunk 6]
```diff
@@ -364,9 +357,7 @@ def passkey_with_output(
         :param script: Path to local umockdev script file
         :type script: str
         :param command: Command that needs to check after authentication of user
-        :type command: str | Any
-        :param expected_output_from_cmd: Output we get from command
-        :type expected_output_from_cmd: str | Any
+        :type command: str = exit 0
```
Comment (pbrezina) [inline sssd_test_framework/utils/authentication.py:360]: Default value should go inti `:param:`, e.g. `Command executed after user is authenticated, defaults to "exit 0"`

**Review by pbrezina — CHANGES_REQUESTED** [review]
Currently, I get

```
** (umockdev-run:3503): ERROR **: 12:16:14.339: ScriptRunner op_write[/dev/hidraw1]: data mismatch; got block '^@^C\xcf^D?^@\x8d^B\xa4^Ahipa.test^BX Ab^I(?\xc8X\xf5Dx^_>\xa2,\xa0^Z\xf5\xd0^E\xf4Lz\xcc\xc8^M?\x94\xb0!\xb4^C\x81\xa2bidX@Q@^@^C\xcf^D\xd6^@\xe6 rj\x9d?,EAX\x99\xdd+a\xe2<^H>*\xf6H^K^NKH\xe3\xfda@\x8b\xc9^]=\x8eR\xc0^I\xa5Y\xf2\xe1%\xf8\xbd^X\x91?\xe0*\x9dx^_HE\xe3Q^@^C\xcf^D\xd6^Agq\xe1dtypejpublic-key^E\xa1bup\xf4^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@' (195 bytes), expected block '^@^C\xcf^D?^@\x8d^B\xa4^Ahipa.test^BX ^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^C\x81\xa2bidX@Q@^@^C\xcf^D\xd6^@\xe6 rj\x9d?,EAX\x99\xdd+a\xe2<^H>*\xf6H^K^NKH\xe3\xfda@\x8b\xc9^]=\x8eR\xc0^I\xa5Y\xf2\xe1%\xf8\xbd^X\x91?\xe0*\x9dx^_HE\xe3Q^@^C\xcf^D\xd6^Agq\xe1dtypejpublic-key^E\xa1bup\xf4^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@'
                                                    /tmp/mh.fs.rollback.rL8XkQEv6: line 7:  3503 Trace/breakpoint trap   (core dumped) LD_PRELOAD=/opt/random.so umockdev-run --device '/tmp/mh.fs.rollback.VeyVHwoTQ' --ioctl '/dev/hidraw1=/tmp/mh.fs.rollback.PfVmVAynI' --script '/dev/hidraw1=/tmp/mh.fs.rollback.1mWHJOQ8N' -- '/tmp/mh.fs.rollback.TF8H4ldzH'
```

What do I need to do to make it work?

File: sssd_test_framework/utils/authentication.py (hunk 6 above)
Comment (madhuriupadhye) [reply sssd_test_framework/utils/authentication.py]: updated.

File: sssd_test_framework/utils/authentication.py [hunk 7]
```diff
@@ -399,42 +411,79 @@ def passkey(self, username: str, *, pin: str | int, device: str, ioctl: str, scr
...
+                eof {{exitmsg "Unexpected end of file" 202}}
             }}
 
-            puts "expect result: Unexpected code path"
-            exit 203
-            """
+            send -- "$command\n"
+
+            expect {{
+                "{expected_output_cmd}" {{exitmsg "Got expected output of the command" 2}}
```
[the remark this answers is not in the document]
Comment (madhuriupadhye) [reply sssd_test_framework/utils/authentication.py]: updated.

**Comment by madhuriupadhye** [comment]
> Currently, I get
> 
> ```
> ** (umockdev-run:3503): ERROR **: 12:16:14.339: ScriptRunner op_write[/dev/hidraw1]: data mismatch; got block '^@^C\xcf^D?^@\x8d^B\xa4^Ahipa.test^BX Ab^I(?\xc8X\xf5Dx^_>\xa2,\xa0^Z\xf5\xd0^E\xf4Lz\xcc\xc8^M?\x94\xb0!\xb4^C\x81\xa2bidX@Q@^@^C\xcf^D\xd6^@\xe6 rj\x9d?,EAX\x99\xdd+a\xe2<^H>*\xf6H^K^NKH\xe3\xfda@\x8b\xc9^]=\x8eR\xc0^I\xa5Y\xf2\xe1%\xf8\xbd^X\x91?\xe0*\x9dx^_HE\xe3Q^@^C\xcf^D\xd6^Agq\xe1dtypejpublic-key^E\xa1bup\xf4^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@' (195 bytes), expected block '^@^C\xcf^D?^@\x8d^B\xa4^Ahipa.test^BX ^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^A^C\x81\xa2bidX@Q@^@^C\xcf^D\xd6^@\xe6 rj\x9d?,EAX\x99\xdd+a\xe2<^H>*\xf6H^K^NKH\xe3\xfda@\x8b\xc9^]=\x8eR\xc0^I\xa5Y\xf2\xe1%\xf8\xbd^X\x91?\xe0*\x9dx^_HE\xe3Q^@^C\xcf^D\xd6^Agq\xe1dtypejpublic-key^E\xa1bup\xf4^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@'
>                                                     /tmp/mh.fs.rollback.rL8XkQEv6: line 7:  3503 Trace/breakpoint trap   (core dumped) LD_PRELOAD=/opt/random.so umockdev-run --device '/tmp/mh.fs.rollback.VeyVHwoTQ' --ioctl '/dev/hidraw1=/tmp/mh.fs.rollback.PfVmVAynI' --script '/dev/hidraw1=/tmp/mh.fs.rollback.1mWHJOQ8N' -- '/tmp/mh.fs.rollback.TF8H4ldzH'
> ```
> 
> What do I need to do to make it work?

you need to add `Environment=LD_PRELOAD=/opt/random.so` under `[service]` to `/usr/lib/systemd/system/ipa-otpd@.service` to ipa server. And then `systemctl daemon-reload` and `ipactl restart`.

**Comment by pbrezina** [comment]
Hi Madhuri, please squash the two commits into a single one.

I think you did a good job, but it is still not yet complete. Currently, it does not return correct return code of the command - it always returns 0. You can test it by passing `false` instead of `klist`. I would like to propose the following change:

```diff
diff --git a/sssd_test_framework/utils/authentication.py b/sssd_test_framework/utils/authentication.py
index 8950dcc..a42f15f 100644
--- a/sssd_test_framework/utils/authentication.py
+++ b/sssd_test_framework/utils/authentication.py
@@ -381,7 +381,7 @@ class SUAuthenticationUtils(MultihostUtility[MultihostHost]):
                 chmod 755 /usr/libexec/sssd/passkey_child
                 chmod -R a+rwx $UMOCKDEV_DIR
 
-                su --shell /bin/sh nobody -c "su - '{username}'"
+                su --shell /bin/sh nobody -c "su - '{username}' -c '{command}'"
                 """,
             mode="a=rx",
         )
@@ -438,24 +438,12 @@ class SUAuthenticationUtils(MultihostUtility[MultihostHost]):
             }}
 
             expect {{
-                -re $prompt {{
-                    if {{ $command eq "exit 0" }} {{ exitmsg "Password authentication successful" 0}}
-                }}
                 "Authentication failure" {{exitmsg "Authentication failure" 1}}
+                eof {{exitmsg "Password authentication successful" 0}}
                 timeout {{exitmsg "Unexpected output" 201}}
-                eof {{exitmsg "Unexpected end of file" 202}}
-            }}
-
-            send -- "$command\n"
-
-            expect {{
-                -re $prompt {{ exitmsg "Get the output of command" 3 }}
-                timeout {{exitmsg "Unexpected output" 201}}
-                eof {{exitmsg "Unexpected end of file" 202}}
             }}
 
             exitmsg "Unexpected code path" 203
-
             """,
             verbose=False,
         )
```

It simplifies it a little bit and you get the correct command return code.

**Comment by madhuriupadhye** [comment]
> Hi Madhuri, please squash the two commits into a single one.
> 
> I think you did a good job, but it is still not yet complete. Currently, it does not return correct return code of the command - it always returns 0. You can test it by passing `false` instead of `klist`. I would like to propose the following change:
> 
> ```diff
> diff --git a/sssd_test_framework/utils/authentication.py b/sssd_test_framework/utils/authentication.py
> index 8950dcc..a42f15f 100644
> --- a/sssd_test_framework/utils/authentication.py
> +++ b/sssd_test_framework/utils/authentication.py
> @@ -381,7 +381,7 @@ class SUAuthenticationUtils(MultihostUtility[MultihostHost]):
>                  chmod 755 /usr/libexec/sssd/passkey_child
>                  chmod -R a+rwx $UMOCKDEV_DIR
>  
> -                su --shell /bin/sh nobody -c "su - '{username}'"
> +                su --shell /bin/sh nobody -c "su - '{username}' -c '{command}'"
>                  """,
>              mode="a=rx",
>          )
> @@ -438,24 +438,12 @@ class SUAuthenticationUtils(MultihostUtility[MultihostHost]):
>              }}
>  
>              expect {{
> -                -re $prompt {{
> -                    if {{ $command eq "exit 0" }} {{ exitmsg "Password authentication successful" 0}}
> -                }}
>                  "Authentication failure" {{exitmsg "Authentication failure" 1}}
> +                eof {{exitmsg "Password authentication successful" 0}}
>                  timeout {{exitmsg "Unexpected output" 201}}
> -                eof {{exitmsg "Unexpected end of file" 202}}
> -            }}
> -
> -            send -- "$command\n"
> -
> -            expect {{
> -                -re $prompt {{ exitmsg "Get the output of command" 3 }}
> -                timeout {{exitmsg "Unexpected output" 201}}
> -                eof {{exitmsg "Unexpected end of file" 202}}
>              }}
>  
>              exitmsg "Unexpected code path" 203
> -
>              """,
>              verbose=False,
>          )
> ```
> 
> It simplifies it a little bit and you get the correct command return code.

Yes, it worked expected.  It's a simpler version which I did not even think. Thanks!

File: requirements.txt (hunk 4 above)
Comment (pbrezina) [reply requirements.txt:4]: Remove this part and we can merge.

**Review by pbrezina — CHANGES_REQUESTED** [review]

**Pull request merged**